### PR TITLE
Load balancer sample using Consul and Least Connection [TCP]

### LINKED ISSUE #1811: The SSL connection could not be established - While setting up LoadBalancer using Consul and LeastConnection.
### Discussed in https://github.com/ThreeMammals/Ocelot/discussions/1804

<div type='discussions-op-text'>

<sup>Originally posted by **ks1990cn** November 24, 2023</sup>
Hi @raman-m @ggnaegi / Anyone ,

Can you please help me to resolve error mentioned in title? 
I've created 3 web api projects - 2 of them to replicate DownstreamApi Instance1 and Instance2, 1 is to replicate API Gateway.

when I try to reach downstream using upstream of API Gateway, and when it tries to find hostandport anc check for services available, it suddenly breaks stating SSL Connection could not be established.

Can you help me by cloning below code and let me know what I am missing in setup. Here is GIF of problem and code from where you can clone.

Please let me know any input required from me.

SSLissue

Code path changes I've done:
https://github.com/ks1990cn/Ocelot/tree/LoadBalancerSample</div>

### PR DESCRIPTION
**Pull request by ks1990cn**
## Closes #1811 
- #1811 

LoadBalancer Sample using Consul and Least Connection type.

I am creating this PR as a handy guide for local setup and running so any contributor can see loadbalancer in action with minimal repeated setup.

I think this can also help me / anyone as always present setup so just quickly go and make required changes to debug any requested changes / bug.

## Proposed Changes

- Under Sample folder, created LoadBalancerDemo
- This folder have 3 Services (Instance1, Instance2 - to demo 2 instances & API Gateway).
- When we hit - /hostname of API Gateway, it loadbalancer will come into picture and with least connection method it will return to which instance it is serving request.
- Ocelot.Sample.Loadbalancer.Common for setup of Consul with Instance 1 and instance 2 (Services)

### CODE REVIEW

**Comment by ks1990cn** [comment]
@raman-m / @ggnaegi / @RaynaldM  I am raising clean PR again. There was lot of changes and some how lot errors I use to get while working on docker on that branch.

Please add required reviewers again.

I am adding more queries on based on this PR @ggnaegi / @raman-m. Please help me when you get chance.

**Comment by ks1990cn** [comment]
@ggnaegi I am trying to understand how I can run multiple instances of an Image using docker compose.
I am facing an issue here, when my instance starts in container - I am able to reach to one port assigned in container but unable to reach to another port which is assigned for same image.

here is my experimental branch where I tried to do this thing. Can you quickly go through my code and check what I am missing in docker-compose or any other thing?
dockerquery2

**Comment by ks1990cn** [comment]
> Tanmay, why did you close #1810 ?

Code and changes got into mess and my solution started giving lot of local errors. so I had to create branch again & solving them was taking more than creating & chery picking again.

**Comment by raman-m** [comment]
You've got the perfect [code review](https://github.com/ThreeMammals/Ocelot/pull/1810#pullrequestreview-1763465150) of #1810 by @ggnaegi 
Seems you have to take into account old code review issues too...

**Comment by ks1990cn** [comment]
> You've got the perfect [code review](https://github.com/ThreeMammals/Ocelot/pull/1810#pullrequestreview-1763465150) of #1810 by @ggnaegi Seems you have to take into account old code review issues too...

Yes I am trying to clean referring to same. but I am struggling with docker instances. I need help here.

**Comment by ks1990cn** [comment]
@ggnaegi Please help me with https://github.com/ThreeMammals/Ocelot/pull/1810#discussion_r1414529390 comment also & how I can set my host name to consul while running on docker? It gets hosted with localhost.

**Comment by ks1990cn** [comment]
> > @ks1990cn [commented](#issuecomment-1847034568)
> 
> Route JSON [ServiceName](https://ocelot.readthedocs.io/en/latest/features/servicediscovery.html#consul-provider-type:~:text=%22ServiceName%22%3A%20%22product%22%2C)

@raman-m example also suggest for host as localhost, but @ggnaegi suggesting for Consul https://github.com/ThreeMammals/Ocelot/pull/1810#discussion_r1414540043

image

Is he want to explain something different?

**Comment by raman-m** [comment]
Wait for our Consul expert, @ggnaegi ! 😉 👑 aka King of the England!

**Review by RaynaldM — APPROVED** [review]

**Comment by ks1990cn** [comment]
> [#1810 (comment)](https://github.com/ThreeMammals/Ocelot/pull/1810#discussion_r1414528706)
> 
> @ggnaegi I am trying to understand how I can run multiple instances of an Image using docker compose. I am facing an issue here, when my instance starts in container - I am able to reach to one port assigned in container but unable to reach to another port which is assigned for same image.
> 
> here is my experimental branch where I tried to do this thing. Can you quickly go through my code and check what I am missing in docker-compose or any other thing?
> 
> dockerquery2 [ dockerquery2 ](https://private-user-images.githubusercontent.com/29522704/288714678-d68021c0-4f7f-4f66-8350-06da8a105bc7.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDIxMTMyMzAsIm5iZiI6MTcwMjExMjkzMCwicGF0aCI6Ii8yOTUyMjcwNC8yODg3MTQ2NzgtZDY4MDIxYzAtNGY3Zi00ZjY2LTgzNTAtMDZkYThhMTA1YmM3LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjA5VDA5MDg1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThkYzM1YzVjY2NmZDEyODRjYzM0ZTEwNjFiOTc5Njk2MGQ0ZWFkMWM5MTQwMDIxMmYxOTIzMDk3ZWRhZWVjYzImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.3LFlmSYVvSYYn7pWX68tIwbU5zD8SIPPNljJW18kZiY) [ ](https://private-user-images.githubusercontent.com/29522704/288714678-d68021c0-4f7f-4f66-8350-06da8a105bc7.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDIxMTMyMzAsIm5iZiI6MTcwMjExMjkzMCwicGF0aCI6Ii8yOTUyMjcwNC8yODg3MTQ2NzgtZDY4MDIxYzAtNGY3Zi00ZjY2LTgzNTAtMDZkYThhMTA1YmM3LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjA5VDA5MDg1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThkYzM1YzVjY2NmZDEyODRjYzM0ZTEwNjFiOTc5Njk2MGQ0ZWFkMWM5MTQwMDIxMmYxOTIzMDk3ZWRhZWVjYzImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.3LFlmSYVvSYYn7pWX68tIwbU5zD8SIPPNljJW18kZiY)

@ggnaegi Here I can understand is first docker image builds and creates port and I am trying to assign another port on same built, which is causing this issue. I am waiting for your reply what kind of approach you are suggesting to use only 1 project.

https://github.com/ThreeMammals/Ocelot/pull/1810#discussion_r1414528706
cc: @raman-m

**Comment by ggnaegi** [comment]
Not even king at home, sometimes in my office, if the dog agrees.

**Comment by raman-m** [comment]
> @ks1990cn commented [on Dec 9](https://github.com/ThreeMammals/Ocelot/pull/1842#issuecomment-1848338074)

Have you ran all your Docker containers for each image?
Seems you need to learn Docker more...
Which Docker software do you use except Visual Studio Docker support?

**Comment by ks1990cn** [comment]
> > @ks1990cn commented [on Dec 9](https://github.com/ThreeMammals/Ocelot/pull/1842#issuecomment-1848338074)
> 
> Have you ran all your Docker containers for each image? Seems you need to learn Docker more... Which Docker software do you use except Visual Studio Docker support?

Same container is running all images which are defined in compose file. 
I use visual studio only, I saw only it as an example in Microsoft tutorials.
Not much expert here, I am learning.

**Comment by raman-m** [comment]
> Not much expert here, I am learning.

Aha! Good luck in learning!

Install [Docker Desktop](https://docs.docker.com/desktop/install/windows-install/) if you use Windows OS...
This is Docker engine management GUI tool. It will allow you to manage Docker artifacts easily...
Also it is possible to use Docker CLI... But it is a bit complicated thing for Docker newbies who learn this software... Just use Docker Desktop UI.
Visual Studio can run one Docker container for current default startup project, but for projects with multiple Docker images you have to run multiple Visual Studio sessions... It is not required if you use Docker Desktop for objects managements!
Docker Compose project just allows to create Docker Network for all Docker images in the solution. They can communicate to each other via Docker Network of composing project. But Compose project doesn't manage all Docker objects of the solution, it complies all Docker filed as images and nothing more. You have to manage Docker objects manually somehow...

I will return to this PR back after fixing the problem of placing all this code into the root repo folder...
So, work on this issue now ❗

**Comment by ks1990cn** [comment]
> > Not much expert here, I am learning.
> 
> Aha! Good luck in learning!
> 
> Install [Docker Desktop](https://docs.docker.com/desktop/install/windows-install/) if you use Windows OS... It installs also Docker engine, CLI, virtual machine aka WLS, and UI app on top of them. This is Docker engine management GUI tool. It will allow you to manage Docker artifacts easily... Also it is possible to use Docker CLI... But it is a bit complicated thing for Docker newbies who learn this software... Just use Docker Desktop UI. Visual Studio can run one Docker container for current default startup project, but for projects with multiple Docker images you have to run multiple Visual Studio sessions... It is not required if you use Docker Desktop for objects managements! Docker Compose project just allows to create Docker Network for all Docker images in the solution. They can communicate to each other via Docker Network of composing project. But Compose project doesn't manage all Docker objects of the solution, it complies all Docker filed as images and nothing more. You have to manage Docker objects manually somehow...
> 
> I will return to this PR back after fixing the problem of placing all this code into the root repo folder... So, work on this issue now ❗

Ah! I have docker desktop, I thought you are asking about tool to write docker code.

**Comment by ggnaegi** [comment]
@ks1990cn I will provide you a sample implementation during the week. Be patient, i'm a bit busy at the minute. Thanks!

**Comment by ks1990cn** [comment]
> Tanmay, you 've created the solution and projects in the root folder! That's wrong!!! We have special folder for all samples: [samples](https://github.com/ThreeMammals/Ocelot/tree/develop/samples)
> 
> You must move solution/projects into a folder inside of the [samples](https://github.com/ThreeMammals/Ocelot/tree/develop/samples) folder!
> 
> image

Made respective changes

**Comment by raman-m** [comment]
@ks1990cn commented [on Dec 12](https://github.com/ThreeMammals/Ocelot/pull/1842#issuecomment-1851489331):
> Made respective changes
> https://github.com/ThreeMammals/Ocelot/pull/1842/commits/b6257233d80370982eeaf7be39d76529a53a7695

Is that your change? Are you kidding?
Please, work on major code review issue aka [root folder problem](https://github.com/ThreeMammals/Ocelot/pull/1842#pullrequestreview-1775006696)
After that, as a team, we will perform classic code review: C#, design etc...

**Comment by raman-m** [comment]
Your feature branch: https://github.com/ks1990cn/Ocelot/tree/lbsample
You've moved **NOTHING**!
image
I have last drop of my patience...

**Comment by raman-m** [comment]
I've [rebased](https://github.com/ThreeMammals/Ocelot/pull/1842#commits-pushed-c53c297) the branch.
Probably you did the fix in local repo.

- Delete local branch in local repo.
- Checkout the feature branch once again
- Apply the fix
- Push

**Comment by ks1990cn** [comment]
> Your feature branch: https://github.com/ks1990cn/Ocelot/tree/lbsample You've moved **NOTHING**! image I have last drop of my patience...

image


> Your feature branch: https://github.com/ks1990cn/Ocelot/tree/lbsample You've moved **NOTHING**! image I have last drop of my patience...

You don't need to loose patience, I am not sure whats wrong, let me figure out.

image

@raman-m Steps you mentioned didn't worked for me. Let me try again, maybe I need to pull through this fork and remove my local folder too.

**Comment by ks1990cn** [comment]
Closing this PR, cause this had folder structure issue. Clean PR is raised again.

**Comment by raman-m** [comment]
Tanmay, there is no need to close a PR each time when the code reviewer found an issue.
You just need update your feature branch! Nothing more! The changes will be reflected to a PR...
This is the 2nd closing... I have no words...

**Comment by raman-m** [comment]
Also, creation of new PR based on old feature branch has no sense, because files in the [feature branch](https://github.com/ks1990cn/Ocelot/tree/lbsample) remain the same!

**Comment by ks1990cn** [comment]
I agree on you @raman-m , but I figured out sample folder structure is little different thing. On File explorer it different and on visual studio naming and structuring is different.

I already had project file into it which was created to wrong path. I had no chance for it. In my new code, I've made some more docker changes which @ggnaegi was suggesting (Just facing another issue with cerficates).

**Comment by ks1990cn** [comment]
Correct
